File: online/LeetCodeCN/1~100/1.ts
// 两数之和
// https://leetcode-cn.com/problems/two-sum/

function twoSum(nums: number[], target: number): number[] {
    const mapNum2Idx = new Map()
    for (let i = 0; i < nums.length; ++i) {
        const complement = target - nums[i]
        if (mapNum2Idx.has(complement)) {
            return [mapNum2Idx.get(complement), i]
        }
        mapNum2Idx.set(nums[i], i)
    }

    return []
};

File: online/LeetCodeCN/1~100/1.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, int> mp;
        for (unsigned i = 0; i < nums.size(); ++i) {
            mp[nums[i]] = i;
        }
        for (unsigned i = 0; i < nums.size(); ++i) {
            int num = nums[i];
            int add = target - num;
            if (mp.find(add) != mp.end() && mp[add] != i) {
                return vector<int> {int(i), mp[add]};
            }
        }
        return {};
    }
};
